File: vertx-kotlin-book/code-go/internal/service/service.go
// Package service is the thin application-service layer. Business logic that
// doesn't belong in the repository (validation orchestration, cross-aggregate
// checks) lives here, keeping it testable without a database.
package service

import (
	"context"

	"github.com/example/users-go/internal/db"
	"github.com/example/users-go/internal/domain"
)

type UserService struct {
	repo *db.Repository
}

func New(repo *db.Repository) *UserService { return &UserService{repo: repo} }

func (s *UserService) GetByID(ctx context.Context, id int64) (domain.User, error) {
	u, err := s.repo.FindByID(ctx, id)
	if err != nil {
		return domain.User{}, err
	}
	if u == nil {
		return domain.User{}, &domain.NotFoundError{ID: id}
	}
	return *u, nil
}

func (s *UserService) Create(ctx context.Context, in domain.NewUser) (domain.User, error) {
	// Friendly uniqueness pre-check so we can return a clean 409 without
	// relying on the DB error; still catch DuplicateEmail from the repo
	// because races exist.
	existing, err := s.repo.FindByEmail(ctx, in.Email)
	if err != nil {
		return domain.User{}, err
	}
	if existing != nil {
		return domain.User{}, &domain.DuplicateEmailError{Email: in.Email}
	}
	return s.repo.Create(ctx, in)
}

func (s *UserService) BulkCreate(ctx context.Context, inputs []domain.NewUser) ([]domain.User, error) {
	return s.repo.CreateMany(ctx, inputs)
}

// StreamAll yields each user to fn; a non-nil fn error stops the stream.
func (s *UserService) StreamAll(ctx context.Context, emailPrefix *string, fn func(domain.User) error) error {
	return s.repo.StreamAll(ctx, emailPrefix, 100, fn)
}

File: vertx-kotlin-book/code-rust/src/service.rs
//! Thin application-service layer: orchestration that doesn't belong in the
//! repository (validation flow, uniqueness pre-check) lives here, keeping it
//! independent of any transport. Mirrors the Go `internal/service`.

use futures::Stream;

use crate::db::Repository;
use crate::domain::{NewUser, User, UserError};

#[derive(Clone)]
pub struct UserService {
    repo: Repository,
}

impl UserService {
    pub fn new(repo: Repository) -> Self {
        Self { repo }
    }

    pub async fn get_by_id(&self, id: i64) -> Result<User, UserError> {
        self.repo
            .find_by_id(id)
            .await?
            .ok_or(UserError::NotFound(id))
    }

    pub async fn create(&self, input: NewUser) -> Result<User, UserError> {
        // Friendly uniqueness pre-check so we can return a clean conflict
        // without relying on the DB error; still catch DuplicateEmail from the
        // repo because races exist.
        if self.repo.find_by_email(&input.email).await?.is_some() {
            return Err(UserError::DuplicateEmail(input.email));
        }
        self.repo.create(input).await
    }

    pub async fn bulk_create(&self, inputs: &[NewUser]) -> Result<Vec<User>, UserError> {
        self.repo.create_many(inputs).await
    }

    /// Streams every user; a slow consumer back-pressures all the way to the
    /// Postgres portal (see `Repository::stream_all`).
    pub fn stream_all(
        &self,
        email_prefix: Option<String>,
    ) -> impl Stream<Item = Result<User, UserError>> {
        self.repo.stream_all(email_prefix)
    }
}
